File: agenkit-go/tests/cross_language_harness.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"
)

const (
	ProtocolVersion = "1.0"
	Version         = "0.41.0"

	// Exit codes
	ExitSuccess         = 0
	ExitError           = 1
	ExitProtocolError   = 2
	ExitTimeout         = 3
	ExitInternalError   = 4
)

// Protocol message structures

type Request struct {
	ProtocolVersion string                 `json:"protocol_version"`
	RequestID       string                 `json:"request_id"`
	Command         string                 `json:"command"`
	Payload         map[string]interface{} `json:"payload"`
}

type Response struct {
	ProtocolVersion string                 `json:"protocol_version"`
	RequestID       string                 `json:"request_id"`
	Status          string                 `json:"status"`
	Result          map[string]interface{} `json:"result,omitempty"`
	Error           *ErrorInfo             `json:"error,omitempty"`
}

type ErrorInfo struct {
	Type       string                 `json:"type"`
	Message    string                 `json:"message"`
	Details    map[string]interface{} `json:"details,omitempty"`
	StackTrace string                 `json:"stack_trace,omitempty"`
}

type TestPayload struct {
	Pattern    string                 `json:"pattern"`
	ScenarioID string                 `json:"scenario_id"`
	Input      map[string]interface{} `json:"input"`
}

type Message struct {
	Role     string                 `json:"role"`
	Content  string                 `json:"content"`
	Metadata map[string]interface{} `json:"metadata"`
}

type BehaviorData struct {
	Turns      int      `json:"turns,omitempty"`
	ToolCalls  []string `json:"tool_calls,omitempty"`
	SubAgents  []string `json:"sub_agents,omitempty"`
	Iterations int      `json:"iterations,omitempty"`
}

type ExecutionInfo struct {
	DurationMs  float64 `json:"duration_ms"`
	LLMCalls    int     `json:"llm_calls,omitempty"`
	TokensUsed  int     `json:"tokens_used,omitempty"`
	MemoryBytes int64   `json:"memory_bytes,omitempty"`
}

// Pattern registry
var supportedPatterns = map[string]bool{
	"reflection":           true,
	"sequential":           true,
	"parallel":             true,
	"router":               true,
	"react":                true,
	"conversational":       true,
	"agents_as_tools":      true,
	"fallback":             true,
	"supervisor":           true,
	"planning":             true,
	"task":                 true,
	"collaborative":        true,
	"human_in_loop":        true,
	"autonomous":           true,
	"multiagent":           true,
	"orchestration":        true,
	"memory":               true,
	"reasoning_with_tools": true,
}

func main() {
	// Read request from stdin
	requestData, err := io.ReadAll(os.Stdin)
	if err != nil {
		writeErrorResponse("", "InternalError", fmt.Sprintf("Failed to read stdin: %v", err))
		os.Exit(ExitInternalError)
	}

	// Parse request
	var request Request
	if err := json.Unmarshal(requestData, &request); err != nil {
		writeErrorResponse("", "ProtocolError", fmt.Sprintf("Invalid JSON: %v", err))
		os.Exit(ExitProtocolError)
	}

	// Handle request
	response := handleRequest(&request)

	// Write response
	responseData, err := json.Marshal(response)
	if err != nil {
		writeErrorResponse(request.RequestID, "InternalError", fmt.Sprintf("Failed to marshal response: %v", err))
		os.Exit(ExitInternalError)
	}

	fmt.Println(string(responseData))

	// Exit with appropriate code
	if response.Status == "success" {
		os.Exit(ExitSuccess)
	}
	os.Exit(ExitError)
}

func handleRequest(request *Request) *Response {
	// Validate protocol version
	if request.ProtocolVersion != ProtocolVersion {
		return &Response{
			ProtocolVersion: ProtocolVersion,
			RequestID:       request.RequestID,
			Status:          "error",
			Error: &ErrorInfo{
				Type:    "ProtocolError",
				Message: fmt.Sprintf("Protocol version mismatch: expected %s, got %s", ProtocolVersion, request.ProtocolVersion),
			},
		}
	}

	// Route command
	var result map[string]interface{}
	var err *ErrorInfo

	switch request.Command {
	case "execute_test":
		result, err = executeTest(request.Payload)
	case "get_info":
		result, err = getInfo()
	case "health_check":
		result, err = healthCheck()
	default:
		err = &ErrorInfo{
			Type:    "CommandNotFound",
			Message: fmt.Sprintf("Unknown command: %s", request.Command),
		}
	}

	// Build response
	response := &Response{
		ProtocolVersion: ProtocolVersion,
		RequestID:       request.RequestID,
	}

	if err != nil {
		response.Status = "error"
		response.Error = err
	} else {
		response.Status = "success"
		response.Result = result
	}

	return response
}

func executeTest(payload map[string]interface{}) (map[string]interface{}, *ErrorInfo) {
	// Parse test payload
	pattern, ok := payload["pattern"].(string)
	if !ok {
		return nil, &ErrorInfo{
			Type:    "ValidationError",
			Message: "Pattern name is required",
		}
	}

	// Normalize pattern name to lowercase for case-insensitive matching
	patternLower := strings.ToLower(pattern)

	_, ok = payload["scenario_id"].(string)
	if !ok {
		return nil, &ErrorInfo{
			Type:    "ValidationError",
			Message: "Scenario ID is required",
		}
	}

	input, ok := payload["input"].(map[string]interface{})
	if !ok {
		return nil, &ErrorInfo{
			Type:    "ValidationError",
			Message: "Input is required",
		}
	}

	// Check if pattern is supported
	if !supportedPatterns[patternLower] {
		return nil, &ErrorInfo{
			Type:    "PatternNotFound",
			Message: fmt.Sprintf("Pattern '%s' not implemented in Go harness", pattern),
		}
	}

	// Parse input message
	messageData, ok := input["message"].(map[string]interface{})
	if !ok {
		return nil, &ErrorInfo{
			Type:    "ValidationError",
			Message: "Input message is required",
		}
	}

	role, _ := messageData["role"].(string)
	content, _ := messageData["content"].(string)
	metadata, _ := messageData["metadata"].(map[string]interface{})

	message := Message{
		Role:     role,
		Content:  content,
		Metadata: metadata,
	}

	// Get configuration
	config, _ := input["config"].(map[string]interface{})

	// Execute pattern
	ctx := context.Background()
	startTime := time.Now()

	result, err := executePattern(ctx, patternLower, message, config)
	if err != nil {
		return nil, &ErrorInfo{
			Type:    "ExecutionError",
			Message: err.Error(),
		}
	}

	duration := time.Since(startTime)

	// Build execution info
	executionInfo := ExecutionInfo{
		DurationMs: float64(duration.Milliseconds()),
		LLMCalls:   0, // TODO: Track actual LLM calls
		TokensUsed: 0, // TODO: Track actual token usage
	}

	// Return result
	return map[string]interface{}{
		"output": map[string]interface{}{
			"message": map[string]interface{}{
				"role":     result.Role,
				"content":  result.Content,
				"metadata": result.Metadata,
			},
			"behavior": map[string]interface{}{
				"turns":      1, // TODO: Track actual turns
				"tool_calls": []string{},
				"sub_agents": []string{},
			},
		},
		"execution_info": map[string]interface{}{
			"duration_ms": executionInfo.DurationMs,
			"llm_calls":   executionInfo.LLMCalls,
			"tokens_used": executionInfo.TokensUsed,
		},
	}, nil
}

func executePattern(ctx context.Context, patternName string, message Message, config map[string]interface{}) (*Message, error) {
	// This is a simplified implementation that returns mock responses
	// TODO: Implement actual pattern execution based on patternName and config

	switch patternName {
	case "reflection":
		return executeReflection(ctx, message, config)
	case "sequential":
		return executeSequential(ctx, message, config)
	case "parallel":
		return executeParallel(ctx, message, config)
	// Add other patterns...
	default:
		// Mock response for now
		return &Message{
			Role:    "assistant",
			Content: fmt.Sprintf("Mock response for %s pattern", patternName),
			Metadata: map[string]interface{}{
				"pattern":   patternName,
				"mock":      true,
			},
		}, nil
	}
}

func executeReflection(ctx context.Context, message Message, config map[string]interface{}) (*Message, error) {
	// TODO: Implement actual reflection pattern execution
	// For now, return a mock response

	maxIterations := 3
	if maxIter, ok := config["max_iterations"].(float64); ok {
		maxIterations = int(maxIter)
	}

	return &Message{
		Role:    "assistant",
		Content: fmt.Sprintf("Reflected response to: %s", message.Content),
		Metadata: map[string]interface{}{
			"iterations": 1,
			"improved":   true,
			"max_iterations": maxIterations,
		},
	}, nil
}

func executeSequential(ctx context.Context, message Message, config map[string]interface{}) (*Message, error) {
	// TODO: Implement actual sequential pattern execution
	agents, _ := config["agents"].([]interface{})
	agentCount := len(agents)

	return &Message{
		Role:    "assistant",
		Content: fmt.Sprintf("Sequential result: %s", message.Content),
		Metadata: map[string]interface{}{
			"agent_count": agentCount,
		},
	}, nil
}

func executeParallel(ctx context.Context, message Message, config map[string]interface{}) (*Message, error) {
	// TODO: Implement actual parallel pattern execution
	agents, _ := config["agents"].([]interface{})
	agentCount := len(agents)

	return &Message{
		Role:    "assistant",
		Content: fmt.Sprintf("Parallel result: %s", message.Content),
		Metadata: map[string]interface{}{
			"agent_count": agentCount,
		},
	}, nil
}

func getInfo() (map[string]interface{}, *ErrorInfo) {
	return map[string]interface{}{
		"language": "go",
		"version":  Version,
		"patterns_supported": []string{
			"reflection",
			"sequential",
			"parallel",
			"router",
			"react",
			"conversational",
			"agents_as_tools",
			"fallback",
			"supervisor",
			"planning",
			"task",
			"collaborative",
			"human_in_loop",
			"autonomous",
			"multiagent",
			"orchestration",
			"memory",
			"reasoning_with_tools",
		},
		"capabilities": map[string]interface{}{
			"streaming":     true,
			"async":         true,
			"llm_providers": []string{"openai", "anthropic"},
		},
	}, nil
}

func healthCheck() (map[string]interface{}, *ErrorInfo) {
	return map[string]interface{}{
		"healthy":        true,
		"uptime_seconds": 0.0, // Stateless harness
	}, nil
}

func writeErrorResponse(requestID, errorType, message string) {
	response := Response{
		ProtocolVersion: ProtocolVersion,
		RequestID:       requestID,
		Status:          "error",
		Error: &ErrorInfo{
			Type:    errorType,
			Message: message,
		},
	}

	data, _ := json.Marshal(response)
	fmt.Println(string(data))
}

File: agenkit-ts/tests/cross_language_harness.ts
#!/usr/bin/env node
/**
 * TypeScript test harness for cross-language equivalence testing.
 *
 * Implements the JSON protocol for executing pattern tests.
 */

const PROTOCOL_VERSION = "1.0";
const VERSION = "0.2.0";

// Exit codes
const EXIT_SUCCESS = 0;
const EXIT_ERROR = 1;
const EXIT_PROTOCOL_ERROR = 2;
const EXIT_TIMEOUT = 3;
const EXIT_INTERNAL_ERROR = 4;

// Protocol message structures
interface Request {
  protocol_version: string;
  request_id: string;
  command: string;
  payload: Record<string, any>;
}

interface ProtocolResponse {
  protocol_version: string;
  request_id: string;
  status: string;
  result?: Record<string, any>;
  error?: ErrorInfo;
}

interface ErrorInfo {
  type: string;
  message: string;
  details?: Record<string, any>;
  stack_trace?: string;
}

interface Message {
  role: string;
  content: string;
  metadata?: Record<string, any>;
}

interface BehaviorData {
  turns?: number;
  tool_calls?: string[];
  sub_agents?: string[];
  iterations?: number;
}

interface ExecutionInfo {
  duration_ms: number;
  llm_calls?: number;
  tokens_used?: number;
  memory_bytes?: number;
}

// Pattern registry
const supportedPatterns: Record<string, boolean> = {
  reflection: true,
  sequential: true,
  parallel: true,
  router: true,
  react: true,
  conversational: true,
  agents_as_tools: true,
  fallback: true,
  supervisor: true,
  planning: true,
  task: true,
  collaborative: true,
  human_in_loop: true,
  autonomous: true,
  multiagent: true,
  orchestration: true,
  memory: true,
  reasoning_with_tools: true,
};

/**
 * Read stdin asynchronously
 */
function readStdin(): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    process.stdin.on('data', (chunk) => chunks.push(chunk));
    process.stdin.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
    process.stdin.on('error', reject);
  });
}

/**
 * Write response to stdout and exit
 */
function writeResponse(response: ProtocolResponse, exitCode: number = EXIT_SUCCESS): never {
  console.log(JSON.stringify(response));
  process.exit(exitCode);
}

/**
 * Write error response and exit
 */
function writeErrorResponse(
  requestId: string,
  errorType: string,
  message: string,
  exitCode: number = EXIT_ERROR
): never {
  const response: ProtocolResponse = {
    protocol_version: PROTOCOL_VERSION,
    request_id: requestId,
    status: 'error',
    error: {
      type: errorType,
      message: message,
    },
  };
  writeResponse(response, exitCode);
}

/**
 * Handle incoming request
 */
function handleRequest(request: Request): ProtocolResponse {
  // Validate protocol version
  if (request.protocol_version !== PROTOCOL_VERSION) {
    return {
      protocol_version: PROTOCOL_VERSION,
      request_id: request.request_id,
      status: 'error',
      error: {
        type: 'ProtocolError',
        message: `Protocol version mismatch: expected ${PROTOCOL_VERSION}, got ${request.protocol_version}`,
      },
    };
  }

  // Route command
  let result: Record<string, any> | undefined;
  let error: ErrorInfo | undefined;

  try {
    switch (request.command) {
      case 'execute_test':
        result = executeTest(request.payload);
        break;
      case 'get_info':
        result = getInfo();
        break;
      case 'health_check':
        result = healthCheck();
        break;
      default:
        error = {
          type: 'CommandNotFound',
          message: `Unknown command: ${request.command}`,
        };
    }
  } catch (e) {
    error = {
      type: e instanceof Error ? e.constructor.name : 'Error',
      message: e instanceof Error ? e.message : String(e),
      stack_trace: e instanceof Error ? e.stack : undefined,
    };
  }

  // Build response
  const response: ProtocolResponse = {
    protocol_version: PROTOCOL_VERSION,
    request_id: request.request_id,
    status: error ? 'error' : 'success',
  };

  if (error) {
    response.error = error;
  } else {
    response.result = result;
  }

  return response;
}

/**
 * Execute a test scenario
 */
function executeTest(payload: Record<string, any>): Record<string, any> {
  // Parse test payload
  const pattern = payload.pattern;
  if (typeof pattern !== 'string') {
    throw new Error('Pattern name is required');
  }

  // Normalize pattern name to lowercase for case-insensitive matching
  const patternLower = pattern.toLowerCase();

  const scenarioId = payload.scenario_id;
  if (typeof scenarioId !== 'string') {
    throw new Error('Scenario ID is required');
  }

  const input = payload.input;
  if (!input || typeof input !== 'object') {
    throw new Error('Input is required');
  }

  // Check if pattern is supported
  if (!supportedPatterns[patternLower]) {
    throw new Error(`Pattern '${pattern}' not implemented in TypeScript harness`);
  }

  // Parse input message
  const messageData = input.message;
  if (!messageData || typeof messageData !== 'object') {
    throw new Error('Input message is required');
  }

  const message: Message = {
    role: messageData.role || 'user',
    content: messageData.content || '',
    metadata: messageData.metadata || {},
  };

  // Get configuration
  const config = input.config || {};

  // Execute pattern
  const startTime = Date.now();
  const outputMessage = executePattern(patternLower, message, config);
  const duration = Date.now() - startTime;

  // Build execution info
  const executionInfo: ExecutionInfo = {
    duration_ms: duration,
    llm_calls: 0, // TODO: Track actual LLM calls
    tokens_used: 0, // TODO: Track actual token usage
  };

  // Return result
  return {
    output: {
      message: {
        role: outputMessage.role,
        content: outputMessage.content,
        metadata: outputMessage.metadata,
      },
      behavior: {
        turns: 1, // TODO: Track actual turns
        tool_calls: [],
        sub_agents: [],
      },
    },
    execution_info: executionInfo,
  };
}

/**
 * Execute a pattern with given message and config
 */
function executePattern(
  patternName: string,
  message: Message,
  config: Record<string, any>
): Message {
  // This is a simplified implementation that returns mock responses
  // TODO: Implement actual pattern execution based on patternName and config

  switch (patternName) {
    case 'reflection':
      return executeReflection(message, config);
    case 'sequential':
      return executeSequential(message, config);
    case 'parallel':
      return executeParallel(message, config);
    default:
      // Mock response for now
      return {
        role: 'assistant',
        content: `Mock response for ${patternName} pattern`,
        metadata: {
          pattern: patternName,
          mock: true,
        },
      };
  }
}

/**
 * Execute reflection pattern
 */
function executeReflection(message: Message, config: Record<string, any>): Message {
  // TODO: Implement actual reflection pattern execution
  // For now, return a mock response

  const maxIterations = config.max_iterations || 3;

  return {
    role: 'assistant',
    content: `Reflected response to: ${message.content}`,
    metadata: {
      iterations: 1,
      improved: true,
      max_iterations: maxIterations,
    },
  };
}

/**
 * Execute sequential pattern
 */
function executeSequential(message: Message, config: Record<string, any>): Message {
  // TODO: Implement actual sequential pattern execution
  const agents = config.agents || [];
  const agentCount = agents.length;

  return {
    role: 'assistant',
    content: `Sequential result: ${message.content}`,
    metadata: {
      agent_count: agentCount,
    },
  };
}

/**
 * Execute parallel pattern
 */
function executeParallel(message: Message, config: Record<string, any>): Message {
  // TODO: Implement actual parallel pattern execution
  const agents = config.agents || [];
  const agentCount = agents.length;

  return {
    role: 'assistant',
    content: `Parallel result: ${message.content}`,
    metadata: {
      agent_count: agentCount,
    },
  };
}

/**
 * Get harness information
 */
function getInfo(): Record<string, any> {
  return {
    language: 'typescript',
    version: VERSION,
    patterns_supported: Object.keys(supportedPatterns),
    capabilities: {
      streaming: true,
      async: true,
      llm_providers: ['openai', 'anthropic'],
    },
  };
}

/**
 * Check harness health
 */
function healthCheck(): Record<string, any> {
  return {
    healthy: true,
    uptime_seconds: 0.0, // Stateless harness
  };
}

/**
 * Main entry point
 */
async function main() {
  try {
    // Read request from stdin
    const requestData = await readStdin();

    // Parse request
    let request: Request;
    try {
      request = JSON.parse(requestData);
    } catch (e) {
      writeErrorResponse(
        '',
        'ProtocolError',
        `Invalid JSON: ${e instanceof Error ? e.message : String(e)}`,
        EXIT_PROTOCOL_ERROR
      );
    }

    // Handle request
    const response = handleRequest(request);

    // Write response
    const exitCode = response.status === 'success' ? EXIT_SUCCESS : EXIT_ERROR;
    writeResponse(response, exitCode);
  } catch (e) {
    writeErrorResponse(
      '',
      'InternalError',
      `Internal error: ${e instanceof Error ? e.message : String(e)}`,
      EXIT_INTERNAL_ERROR
    );
  }
}

// Run main
main();
